### PR TITLE
Sharing to snapchat on iOS through CreativeKit

### PR DESCRIPTION
**Pull request by VincentRoest**
# Overview

Well, sharing to Snapchat was a key necessity for my app, so I decided I'd try my luck. Sadly, we have to depend on SnapSDK's CreativeKit. This also requires that production apps need to be reviewed. 

I realised my iOS development skills were a bit rusty, and that Snap's documentation is not ideal during the process. Nonetheless, I got this hacky version working for sharing videos, images, stickers and text content to Snapchat from iOS. The code is not very clean yet, and we could definitely improve it. Especially the "sleep" could use improvement, but I couldn't find a clean way of preventing the `_scSdkSnapApi` to be released before the share was complete during my late night struggle. All suggestions/edits are very welcome.

# Test Plan

1. Create a developer account for Snapchat here https://kit.snapchat.com/portal/, and create a new app there. 
2. Select Creative kit as a kit you will use in your app
3. Add (or modify) the following keys to your project's Info.plist **SCSDKClientId** (I put in the value for my app's "CLIENT ID FOR NON-CONFIDENTIAL OAUTH2 FLOW"), **LSApplicationQueriesSchemes** where you add 'snapchat' in the array just like with whatsapp or facebook. Optional: **SCSDKRedirectUrl**, some redirect URL, like: my-app://. If you add this, **CFBundleURLSchemes** must also contain your redirect URL’s scheme. More info here: https://docs.snapchat.com/docs/creative-kit
4. add `pod 'SnapSDK', :subspecs => ['SCSDKCreativeKit']` to your project's PodFile. @ _maintainers_, could this be optionally included inside the react-native-share project?

# Example code:
- I'm using the "title" options property as the Snap's text component
- I'm using the "attachmentUrl" for attaching a url to the snap.
```
import Share, {ShareSheet}  from 'react-native-share';
import RNFetchBlob from 'rn-fetch-blob'

shareToSnapchat = async (
    mediaUrl = "https://sample-videos.com/video123/mp4/480/big_buck_bunny_480p_1mb.mp4",
  ) => {
    try {
      RNFetchBlob.config({
        fileCache : true,
        appendExt : 'mp4'
      }).fetch('GET', mediaUrl, {})
      .then((res) => {
        let shareOptions = {
          title: "SNAP TEXT CONTENT",
          url: 'file://' + res.path(),
          type: 'video/mp4', // change type here accordingly
          attachmentUrl: 'https://snapchat.com',
          social: "snapchat",
        }
        Share.shareSingle(shareOptions)
          .then((res) => {
            // undefined for now.
            console.log('res:', res)
          })
          .catch((err) => console.log('err', err))
        });
    } catch(e) {
      console.warn(e)
    }
  };
```
# Demo Video of the code above:
example

Some potential improvements (that are not necessary per se):
- [ ] I Did not test base64
- [ ] SnapSDK dependency 
- [ ] Stickers on Image and Video content, and Standalone
- [ ] Cleaner Objective-C code
- [ ] Clever way of releasing / initialising SnapSDK

### CODE REVIEW

**Comment by VincentRoest** [comment]
Hey @jgcmarins! I was under the impression that snapchat sharing was already functional on Android no? I haven't got a chance to see for myself. Otherwise, how would you recommend I'd get started? All android packages seem to be non-native.

**Comment by jgcmarins** [comment]
We're good to go with Android existing implementation.
Can we add more info to README about iOS?
And also add an example for both Android and iOS?

**Comment by VincentRoest** [comment]
I'll try to push the example tomorrow and include a section in the readme for iOS. Do you know of any way how we might make this an optional thing instead of a mandatory dependency?

**Comment by jgcmarins** [comment]
> I'll try to push the example tomorrow and include a section in the readme for iOS.

Nice!

> Do you know of any way how we might make this an optional thing instead of a mandatory dependency?

Didn't figure out. Maybe we gonna have to release a major bump.

**Comment by Andrfas** [comment]
After building an app with this branch - I got an error
`dyld: Library not loaded: @rpath/PINCache.framework/PINCache
  Referenced from: /Users/andrew.mikhatskiy/Library/Developer/CoreSimulator/Devices/A67AEA68-7434-4B77-B259-CEF24F8D2F04/data/Containers/Bundle/Application/B09D6537-A239-4252-8EEE-05B8949C4FC0/WeAreFriends.app/Frameworks/SCSDKBitmojiKit.framework/SCSDKBitmojiKit
  Reason: image not found`
Do you know the reason why this happends?

**Comment by jgcmarins** [comment]
@Andrfas thanks!
So it's ok on iOS?

**Comment by MateusAndrade** [comment]
@VincentRoest you can update your branch with the last changes on master?

File: ios/RNShare.m [hunk 1]
```diff
@@ -190,7 +208,7 @@ - (NSDictionary *)constantsToExport
     shareController.completionWithItemsHandler = ^(NSString *activityType, BOOL completed, __unused NSArray *returnedItems, NSError *activityError) {
         if (activityError) {
             failureCallback(activityError);
-        } else {
+        } else if (completed || activityType == nil) {
```
Comment (jgcmarins) [inline ios/RNShare.m]: could this be a fix for https://github.com/react-native-community/react-native-share/issues/624?
cc @felippepuhle

**Comment by VincentRoest** [comment]
For a personal project I've also added IG stories, but these can be removed if we want to move with merging this pull request!

**Comment by bingchao-mai** [comment]
Hi, I have a problem in creating a new app for snapchat. 
I do not know what is the Redirect URLs. It seems useless but i can not skip it.

**Comment by thehappydinoa** [comment]
@VincentRoest Did anything ever come of this? Like did you consider making a library for this bc I am trying to do something like YOLO in react native and would love some pointers

**Comment by VincentRoest** [comment]
@bingchao-mai the url is just a bundle url, see this link for a bit more clarity: https://github.com/Snap-Kit/bitmoji-sample/issues/3

@thehappydinoa I haven’t really pursued this further as this was sufficient for my use case. What specific questions do you have? Maybe I can help you out?

**Comment by thehappydinoa** [comment]
@VincentRoest I was hoping to set up some type of react-native library to integrate Login Kit and Creative Kit from Snapchat, I have been looking all over and have found some old ones, but I really like what you did here

**Comment by MateusAndrade** [comment]
> @VincentRoest I was hoping to set up some type of react-native library to integrate Login Kit and Creative Kit from Snapchat, I have been looking all over and have found some old ones, but I really like what you did here

We can follow with that, we just need to ensure that this is currently working fine with our latest changes. I would be glad to help releasing this. :)

**Comment by VincentRoest** [comment]
Looks like RNFetchBlob is having a problem that I didn't really fancy fixing (https://github.com/joltup/rn-fetch-blob/issues/183) so I used a local file in the ios Bundle for testing (needs to be less than 10 secs), but it still works! Don't forget to add the SCSDKClientId to test. 

Covering the LoginKit and CreativeKit should be fairly similar!

**Comment by thehappydinoa** [comment]
Great work man, does this support transparent images for overlay?

**Comment by VincentRoest** [comment]
According to the Snap Kit docs they support any image as sticker (SCSDKSnapSticker at https://docs.snapchat.com/docs/api/ios/), so yeah. I haven’t made the sticker option available but could be done with a simple “if sticker” statement!

**Comment by kilimchoi** [comment]
@VincentRoest can we merge this?

**Comment by MateusAndrade** [comment]
> can we merge this?

Unfortunately, this still broken on Android at the CI. @VincentRoest do you still interest in that?

**Comment by friederbluemle** [comment]
@VincentRoest - It looks like you accidentally added a reference to your personal fork of `react-native-share` in `example/package.json` (using ssh protocol). Can you please undo this (including the 15,000 changed lines in yarn.lock). Thanks!

**Comment by NishidaRyu416** [comment]
@MateusAndrade I think you should not have asked @VincentRoest to add the example of this change before merging the change as the example is unlikely to work without the change.

**Pull request closed without merging**